router/transformer/transformer_test.go

```diff
@@ -0,0 +1,323 @@
+
+func TestProxyRequest(t *testing.T) {
+	t.Setenv("RSERVER_HTTP_CLIENT_TIMEOUT", "1s")
+	Initialization()
+
+	srvMux := mux.NewRouter()
+	srvMux.HandleFunc("/v0/destinations/{destName}/proxy", proxyHandlerFunc)
+	srv := httptest.NewServer(srvMux)
+	defer srv.Close()
+
+	for destName, tc := range proxyResponseTcs {
+		// skip tests for the mentioned destinations
+		if destName == "not_found_dest" {
+			continue
+		}
+		t.Run(tc.name, func(t *testing.T) {
+			tr := rtTf.NewTransformer()
+			// Logic for executing test-cases not manipulating test-cases
+			if tc.rtTimeout.Milliseconds() > 0 {
+				tr.Setup(tc.rtTimeout)
+			} else {
+				// Just a default value
+				tr.Setup(2 * time.Millisecond)
+			}
+			// Logic to include context timing out
+			ctx := context.TODO()
+			var cancelFunc context.CancelFunc
+			if tc.context.Timeout.Milliseconds() > 0 {
+				ctx, cancelFunc = context.WithTimeout(context.TODO(), tc.context.Timeout)
+				defer cancelFunc()
+			} else if tc.context.Cancel {
+				ctx, cancelFunc = context.WithCancel(context.TODO())
+				cancelFunc()
+			}
+
+			reqParams := &rtTf.ProxyRequestParams{
+				ResponseData: tc.postParametersT,
+				DestName:     destName,
+				BaseUrl:      srv.URL,
+			}
+			stCd, resp, contentType := tr.ProxyRequest(ctx, reqParams)
+
+			assert.Equal(t, tc.Expected.StCode, stCd)
+			require.Equal(t, tc.Expected.ContentType, contentType)
+			if gjson.GetBytes([]byte(resp), "message").Raw != "" {
+				require.JSONEq(t, tc.Expected.Body, resp)
+			} else if tc.Expected.StCode == http.StatusGatewayTimeout {
+				expectedBodyStr := fmt.Sprintf(tc.Expected.Body, srv.URL)
+				require.Equal(t, expectedBodyStr, resp)
+			}
+		})
+	}
+
+	// Not found case
+	tc := proxyResponseTcs["not_found_dest"]
+	t.Run(tc.name, func(t *testing.T) {
+		tr := rtTf.NewTransformer()
+		tr.Setup(tc.rtTimeout)
+		ctx := context.TODO()
+		reqParams := &rtTf.ProxyRequestParams{
+			ResponseData: tc.postParametersT,
+			DestName:     "not_found_dest",
+			BaseUrl:      srv.URL,
+		}
+		stCd, resp, contentType := tr.ProxyRequest(ctx, reqParams)
+		assert.Equal(t, tc.Expected.StCode, stCd)
+		require.Equal(t, tc.Expected.ContentType, contentType)
+		expectedBodyStr := fmt.Sprintf(tc.Expected.Body, srv.URL)
+		require.Equal(t, expectedBodyStr, resp)
+	})
+}
+
+// A kind of mock for transformer proxy endpoint in transformer
+func proxyHandlerFunc(w http.ResponseWriter, req *http.Request) {
+	vars := mux.Vars(req)
+	destName, ok := vars["destName"]
+	if !ok {
+		// This case wouldn't occur I guess
+		http.Error(w, "Wrong url being sent", http.StatusInternalServerError)
+		return
+	}
+	tc := proxyResponseTcs[destName]
+	// sleep is being used to mimic the waiting in actual transformer response
+	if tc.Proxy.TimeoutValue.Milliseconds() > 0 {
+		time.Sleep(tc.Proxy.TimeoutValue)
+	}
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(tc.Proxy.StCode)
+	// Lint error fix
+	checkAndSendResponse(&w, []byte(tc.Proxy.Response))
+}
```

Comment:
I would suggest a small change here:
1. to avoid `proxyResponseTcs` need to be a global variable, you can define it inside the test now
2. no need for test cases to be a map
3. improve test isolation by creating a server only to handle a specific test case.

```suggestion
func mockProxy(tc) http.Handler {
	srvMux := mux.NewRouter()
	srvMux.HandleFunc("/v0/destinations/" + tc.destName "/proxy", func() { 
	vars := mux.Vars(req)
	destName, ok := vars["destName"]
	if !ok {
		// This case wouldn't occur I guess
		http.Error(w, "Wrong url being sent", http.StatusInternalServerError)
		return
	}
	// sleep is being used to mimic the waiting in actual transformer response
	if tc.Proxy.TimeoutValue.Milliseconds() > 0 {
		time.Sleep(tc.Proxy.TimeoutValue)
	}
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(tc.Proxy.StCode)
	// Lint error fix
	checkAndSendResponse(&w, []byte(tc.Proxy.Response))

}
```

Reply:
@lvrach can you once check and confirm if the changes are satisfactory ?